### PR TITLE
Adding support for Close Message

### PR DESCRIPTION
**Pull request by mikaelm12**
Adding support for more message types. 
This is basic support for the close message.

### CODE REVIEW

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java [hunk 1]
```diff
@@ -68,6 +68,10 @@ public HubConnection(String url, Transport transport, Logger logger){
                     case STREAM_INVOCATION:
                     case STREAM_ITEM:
                     case CLOSE:
+                        logger.log(LogLevel.Information, "Close message received from server.");
+                        CloseMessage closeMessage = (CloseMessage)message;
+                        stop(closeMessage.getError());
```
[the remark this answers is not in the document]
Comment (mikaelm12) [reply clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java]: No, it just passes the error message into the stop method

[the remark this answers is not in the document]
Comment (mikaelm12) [reply clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java]: Just passing string errors until I add the `HubException`

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java [hunk 2]
```diff
@@ -68,6 +68,10 @@ public HubConnection(String url, Transport transport, Logger logger){
                     case STREAM_INVOCATION:
                     case STREAM_ITEM:
                     case CLOSE:
```
Comment (mikaelm12) [inline clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java]: As written the Close logic runs for both  `STREAM_INVOCATION` and `STREAM_ITEM` message types. I need to move those two cases with the other unsupported message types.

Comment (mikaelm12) [reply clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java]: And a note to myself to add tests that verify exceptions are thrown when unsupported message types are received. Those kind of tests would have caught this

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (davidfowl) [reply clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java]: Is there a closed event on the client?

[the remark this answers is not in the document]
Comment (mikaelm12) [reply clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java]: Not yet

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java [hunk 3]
```diff
@@ -145,16 +149,29 @@ public void start() throws Exception {
         logger.log(LogLevel.Information, "HubConnected started");
     }
 
+
     /**
      * Stops a connection to the server.
      */
-    public void stop(){
-        logger.log(LogLevel.Debug, "Stopping HubConnection");
+    public void stop(String errorMessage) {
```
Comment (anurse) [inline clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java]: The other clients don't allow the user to stop with an error message, so I don't think we should. The version with an error message should be `private`.

File: clients/java/signalr/src/test/java/HubConnectionTest.java [hunk 4]
```diff
@@ -22,6 +22,22 @@ public void checkHubConnectionState() throws Exception {
         assertEquals(HubConnectionState.DISCONNECTED, hubConnection.getConnectionState());
     }
 
+    @Test
+    public void HubConnectionClosesafterCloseMessage() throws Exception {
```
Comment (anurse) [inline clients/java/signalr/src/test/java/HubConnectionTest.java]: nit: HubConnectionCloses**A**fterCloseMessage

File: clients/java/signalr/src/test/java/JsonHubProtocolTest.java [hunk 5]
```diff
@@ -45,6 +45,38 @@ public void ParsePingMessage() {
...
+    public void ParseCloseMessage() {
+        String stringifiedMessage = "{\"type\":7}\u001E";
+        HubMessage[] messages = jsonHubProtocol.parseMessages(stringifiedMessage);
+
+        //We know it's only one message
+        assertEquals(1, messages.length);
+
+        assertEquals(HubMessageType.CLOSE, messages[0].getMessageType());
+
+        //We can safely cast here because we know that it's an invocation message.
```
Comment (anurse) [inline clients/java/signalr/src/test/java/JsonHubProtocolTest.java]: nit: it's a~~n invocation~~ **close** message.

File: clients/java/signalr/src/test/java/JsonHubProtocolTest.java [hunk 6]
```diff
@@ -45,6 +45,38 @@ public void ParsePingMessage() {
...
+    public void ParseCloseMessageWithError() {
+        String stringifiedMessage = "{\"type\":7,\"error\": \"There was an error\"}\u001E";
+        HubMessage[] messages = jsonHubProtocol.parseMessages(stringifiedMessage);
+
+        //We know it's only one message
+        assertEquals(1, messages.length);
+
+        assertEquals(HubMessageType.CLOSE, messages[0].getMessageType());
+
+        //We can safely cast here because we know that it's an invocation message.
```
Comment (anurse) [inline clients/java/signalr/src/test/java/JsonHubProtocolTest.java]: see above :)

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java [hunk 7]
```diff
@@ -145,16 +149,29 @@ public void start() throws Exception {
         logger.log(LogLevel.Information, "HubConnected started");
     }
 
+
```
Comment (BrennanConroy) [inline clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java]: nit: remove newline

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java [hunk 8]
```diff
@@ -145,16 +149,29 @@ public void start() throws Exception {
...
     /**
      * Stops a connection to the server.
      */
-    public void stop(){
-        logger.log(LogLevel.Debug, "Stopping HubConnection");
+    private void stop(String errorMessage) {
+        if(errorMessage != null){
+            logger.log(LogLevel.Error , "HubConnection disconnected with an error %s", errorMessage);
```
Comment (BrennanConroy) [inline clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java]: full stop

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java [hunk 9]
```diff
@@ -145,16 +149,29 @@ public void start() throws Exception {
...
+        } else {
+            logger.log(LogLevel.Debug, "Stopping HubConnection");
```
Comment (BrennanConroy) [inline clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java]: full stop

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java [hunk 10]
```diff
@@ -145,16 +149,29 @@ public void start() throws Exception {
...
+        }
+
         transport.stop();
         connectionState = HubConnectionState.DISCONNECTED;
         logger.log(LogLevel.Information, "HubConnection stopped");
```
Comment (BrennanConroy) [inline clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java]: full stop

**Pull request merged**